Improve translation errors

lehins opened the pull request:
While looking at #2825 and #2836 I realized that TranslationError was a new type in Babbage and thus we can in fact improve it before the release, so the node can use it to produce better errors to the users. I started at the point of the #2825 PR, since it seemed like a sensible cleanup.

While working on improving the type I found a few problems:

* All error constructors could use with some extra information about the piece of transaction that is at fault for the error. The most common information was about problems with TxOuts, but in order to indicate which TxOut is the offender I moved as well as renamed and improved the `TxOutSource`  (previously `OutputSource`). This will be enough to point at the offender.
* `TranslationLogicErrorDoubleDatum` was an error that could never happen in practice, but it was impossible to avoid without some changes to the way inline datums are being extracted from TxOut in a general way. Thus is the addition of `getTxOutDatum` and removal of less powerful `HasField "datum"` instances.


Two more problems found, that ~haven't~ have been resolved ~yet~:

1. [x] In the PR #2825 we remove `basicValidation`, which is redundant for Babbage, however for Alonzo txInfo check does not verify presence of inputs in the utxo, it simply filters them out. Will now produce `TranslationLogicMissingInput`
2. [x] Test `"attempt calculate ExUnits with invalid tx"` is totally wrong.

KtorZ committed:
Remove unused 'CorruptCostModel' constructor.

KtorZ committed:
Remove 'BasicFailure', made redundant by TranslationError/TranslationLogicErrorInput

  The 'BasicFailure' was initially introduced as an extra sanity check to make
  sure that inputs referenced in the transaction were also present in the
  provided UTxO.

  Now however, this is done regardless in the `txInfoIn` function which is used
  to construct the 'PV1.TxInInfo' for both Alonzo & Babbage transactions:

  ```hs
  case Map.lookup txin mp of
    Nothing -> Left TranslationLogicErrorInput
    Just txout -> case transTxOutAddr txout of
  ```

  Thus, the `TranslationLogicErrorInput` branch from `TranslationError` makes
  the `UnknownInputs` branch redundant, and consequently, `BasicFailure`
  redundant completely. To be 100% equivalent, `TranslationLogicErrorInput`
  should be modified to also contain the missing inputs. I didn't do to that
  extend here because `TranslationError` seems to be use as a kind of _enum_,
  with only 0-ary constructors.

lehins committed:
Improve TranslationError by adding offending bots from the Tx to the error

lehins committed:
Get rid of `HasField "datum"` instance in favor of more powerful getTxOutDatum. Avoid the impossible state in TxInfo where both DataHash and inline datum are supplied in TxOut

JaredCorduan commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:
```diff
@@ -104,44 +105,66 @@ import Prettyprinter (Pretty (..))
+  | InlineDatumsNotSupported !(TxOutSource crypto)
+  | ReferenceScriptsNotSupported !(TxOutSource crypto)
+  | ReferenceInputsNotSupported !(Set (TxIn crypto))
+  | TimeTranslationPastHorizon !Text
```
at least `LanguageNotSupported` and `TimeTranslationPastHorizon` are used in the Alonzo era, so we can't change the serialization.

lehins replied:
I pretty sure this is an incorrect statement. Both of these type constructors where introduced after 1.34.1, which is the latest released version of node.

In 1.34.1:

* There was no costmodel for PlutusV2, so `collectTwoPhaseScriptInputs` returned `NoCostModel` error if anyone where to supply PlutusV2 (whichprobably wasn't possible either): https://github.com/input-output-hk/cardano-ledger/blob/1a9ec4ae9e0b09d54e49b2a40c4ead37edadcce5/eras/alonzo/impl/src/Cardano/Ledger/Alonzo/PlutusScriptApi.hs#L173-L177
In Alonzo `collectTwoPhaseScriptInputs` did not rely on txInfo for validation, because it did not fail at all: https://github.com/input-output-hk/cardano-ledger/blob/1a9ec4ae9e0b09d54e49b2a40c4ead37edadcce5/eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs#L351-L358

* And `TimeTranslationErrorPasHorizon` was not even a thing until 2 weeks ago, because `epochInfoSlotToUTCTime` didn't use to fail until this PR: https://github.com/input-output-hk/cardano-base/pull/277

lehins replied:
So, I am pretty confident we are free to change this serialization of this type as we wish before the next release, which is the primary reason I went on fixing it in the first place :smile:

JaredCorduan replied:
indeed you are right, thanks!

lehins committed:
Ensure txInfo in Alonzo validates presence of inputs

This is necessary because `basicValidation` has been removed in #2825

lehins committed:
Fix `"attempt calculate ExUnits with invalid tx"` test

Previously it was simply checking if the txin was not present in an
empty UTxO. Which was more of a test for `txInfo`, rather than
`evaluateTransactionExecutionUnits`

Jimbo4350 commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Tools.hs:
```diff
@@ -150,25 +122,19 @@ evaluateTransactionExecutionUnits ::
   SystemStart ->
   -- | The array of cost models, indexed by the supported languages.
   Array Language CostModel ->
-  -- | If the transaction meets basic validation, we return a map from
```
:tada:

Jimbo4350 commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:
```diff
@@ -104,44 +105,69 @@ import Prettyprinter (Pretty (..))
+      dec n = Invalid n
+
+data TranslationError crypto
+  = ByronTxOutInContext !(TxOutSource crypto)
```
:+1:

Jimbo4350 reviewed (commented):
Nice! Thanks for improving the errors.

lehins committed:
Remove a test that is no longer valid

JaredCorduan commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Data.hs:
```diff
@@ -180,6 +182,37 @@ instance (CC.Crypto c) => HeapWords (StrictMaybe (DataHash c)) where
+
+-- | Datum can be described by a either a data hash or binary data, but not
+-- both. It can also be neither one of them.
+data Datum era
```
much better place :raised_hands:

JaredCorduan reviewed (approved):
Thank you @lehins for improving these errors, this is extremely helpful. The changes all look good to me (though it was a bit tricky makings sure we didn't change the Alonzo semantics). We review the subtle parts of this PR in a call, which I went back and confirmed to myself.

lehins committed:
Prevent Babbage failures from appearing in Alonzo

JaredCorduan commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxos.hs:
```diff
@@ -153,6 +172,45 @@ utxosTransition =
+          failBecause (CollectErrors alonzoFailures)
+      | otherwise -> pure ()
+  where
+    -- BadTranslation was introduced in Babbage, thus we need to filter those failures out.
```
:pray:

JaredCorduan commented:
@lehins the "failed" CI job is yet another instance of the tests actually passing after a re-run, but not reporting properly. With your :+1: I'd like to merge this great PR :)
